src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,173 @@
+            () => ProjectInfoUpdatedImpl(projectKey, projectInfo),
+            cancellationToken
+        ).ConfigureAwait(false);
+    }
+
+    private void ProjectInfoUpdatedImpl(ProjectKey projectKey, RazorProjectInfo? projectInfo)
+    {
+        _projectSnapshotManagerDispatcher.AssertRunningOnDispatcher();
+
+        if (projectInfo is null)
+        {
+            if (_projectSet.Remove(projectKey))
+            {
+                // projectInfo null is "remove"
+                EnqueueUpdateProject(projectKey, projectInfo: null);
+            }
+            else
+            {
+                _logger.LogWarning("Found no existing project key '{0}' but projectInfo is null. Assuming no-op deletion.", projectKey.Id);
+            }
+
+            return;
+        }
+
+        if (!_projectSet.Contains(projectKey))
+        {
+            var configurationFilePath = FilePathNormalizer.Normalize(projectInfo.SerializedFilePath);
+            _logger.LogInformation("Found no existing project key for project key '{0}'. Assuming new project.", projectKey.Id);
+
+            AddProject(configurationFilePath, projectInfo);
+            return;
+        }
+
+        _logger.LogInformation("Project info changed for project '{0}'", projectKey.Id);
+
+        EnqueueUpdateProject(projectKey, projectInfo);
+    }
+
+    private void AddProject(string configurationFilePath, RazorProjectInfo projectInfo)
+    {
+        var projectFilePath = FilePathNormalizer.Normalize(projectInfo.FilePath);
+        var intermediateOutputPath = Path.GetDirectoryName(configurationFilePath).AssumeNotNull();
+        var rootNamespace = projectInfo.RootNamespace;
+
+        var projectKey = _projectService.AddProject(projectFilePath, intermediateOutputPath, projectInfo.Configuration, rootNamespace, projectInfo.DisplayName);
+        _projectSet.Add(projectKey);
+
+        _logger.LogInformation("Project configuration file added for project '{0}': '{1}'", projectFilePath, configurationFilePath);
+        EnqueueUpdateProject(projectKey, projectInfo);
+    }
+
+    private void UpdateProject(ProjectKey projectKey, RazorProjectInfo? projectInfo)
+    {
+        if (projectInfo is null)
+        {
+            ResetProject(projectKey);
+            return;
+        }
+
+        _logger.LogInformation("Actually updating {project} with a real projectInfo", projectKey);
+
+        var projectWorkspaceState = projectInfo.ProjectWorkspaceState ?? ProjectWorkspaceState.Default;
+        var documents = projectInfo.Documents;
+        _projectService.UpdateProject(
+            projectKey,
+            projectInfo.Configuration,
+            projectInfo.RootNamespace,
+            projectInfo.DisplayName,
+            projectWorkspaceState,
+            documents);
+    }
+
+    private async Task UpdateAfterDelayAsync(ProjectKey projectKey)
+    {
+        await Task.Delay(EnqueueDelay).ConfigureAwait(true);
+
+        var delayedProjectInfo = ProjectInfoMap[projectKey];
```

Comment:
I think the configureaway(true) does this, but it would be good to assert that we're still on the dispatcher thread here. Perhaps even a comment so future devs notice the `true` above.